[improve][doc][do-not-merge] single source of truth for functions CLI and add a table for yml configs (#15389)
apache/pulsar: site2/docs/functions-cli.md

---
id: functions-cli
title: Pulsar Functions CLI and YAML configs
sidebar_label: "CLI and YAML configs"
---

## Pulsar admin CLI for functions

> The Pulsar admin interface enables you to create and manage Pulsar Functions through CLI. For the latest and complete information, including commands, flags, and descriptions, refer to [Pulsar admin CLI](https://pulsar.apache.org/tools/pulsar-admin/).

Reviewer: ```suggestion
The Pulsar admin interface enables you to create and manage Pulsar Functions through CLI. For the latest and complete information, including commands, flags, and descriptions, refer to [Pulsar admin CLI](/tools/pulsar-admin/).
```

## YAML configurations for functions

> You can configure a function by using a predefined a .yaml file. The following table outlines the required fields and arguments.

Reviewer: ```suggestion
You can configure a function by using a predefined YAML file. The following table outlines the required fields and arguments.
```

> | Field Name           | Nested Field                       | Type                       | Related Command Argument   | Description                                                                                                                                                                                                                                      |

Reviewer 2: I suggest to put the Nested Field to the last column

Author: Nested fields have been moved to sub-tables for a clear view. PTAL.

|----------------------|------------------------------------|----------------------------|----------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | runtimeFlags         | N/A                                | String                     | N/A                        | Any flags that you want to pass to a runtime. (for process & Kubernetes runtime only)                                                                                                                                                            |

Reviewer: ```suggestion
| runtimeFlags         | N/A                                | String                     | N/A                        | Any flags that you want to pass to a runtime (for process & Kubernetes runtime only).                                                                                                                                                            |
```

| outputSerdeClassName | N/A                                | String                     | `--output-serde-classname` | The SerDe class used for message outputs.                                                                                                                                                                                                        |
| logTopic             | N/A                                | String                     | `--log-topic`              | The topic that the logs of a function are produced to.                                                                                                                                                                                           |
| processingGuarantees | N/A| String | `--processing-guarantees` | The processing guarantees (delivery semantics) applied to a function. Available values: `ATLEAST_ONCE`, `ATMOST_ONCE`, `EFFECTIVELY_ONCE`.
| retainOrdering | N/A| Boolean | `--retain-ordering`	 | Whether functions consume and process messages in order or not. |
| retainKeyOrdering | N/A| Boolean |	`--retain-key-ordering`	| Whether functions consume and process messages in key order or not. |

> | batchBuilder | N/A| String | `--batch-builder` | Use `producerConfig.batchBuilder` instead (will be deprecated in code soon). |

Reviewer: ```suggestion
| batchBuilder | N/A| String | `--batch-builder` | Use `producerConfig.batchBuilder` instead. <br />**Note**: `batchBuilder` will be deprecated in code soon). |
```
do we need to highlight it?

| | ram | Long | `--ram` | The RAM in bytes that need to be allocated per function instance (for process/Kubernetes runtime only). |
| | disk | Long | `--disk` | The disk in bytes that need to be allocated per function instance (for Kubernetes runtime only). |
| fqfn | N/A | String | `--fqfn` | The Fully Qualified Function Name (FQFN) of a function. |
| windowConfig | | | WindowConfig | N/A| N/A|		
| | windowLengthCount | Int | `--window-length-count` | The number of messages per window. |
| | windowLengthDurationMs | Long | `--window-length-duration-ms` | The time duration (in milliseconds) per window. |
| | slidingIntervalCount | Int | `--sliding-interval-count` | The number of messages after which a window slides. |
| | slidingIntervalDurationMs | Long | `--sliding-interval-duration-ms` | The time duration after which a window slides. |
| | lateDataTopic | String | N/A	| N/A |
| | maxLagMs | Long | N/A	| N/A |
| | watermarkEmitIntervalMs | Long | N/A	| N/A |
| | timestampExtractorClassName | String | N/A	 | N/A|
| | actualWindowFunctionClassName | String | N/A	| N/A |

> | timeoutMs | N/A| Long | `--timeout-ms` | The message timeout in milliseconds. |

Reviewer: ```suggestion
| timeoutMs | N/A| Long | `--timeout-ms` | The message timeout (in milliseconds). |
```
keep consistent with previous texts

| jar | N/A| String | `--jar` | The path of the JAR file for a function (if the function is written in Java). It also supports URL paths that workers can download the package from, including HTTP, HTTPS, file (file protocol assuming that file already exists on worker host), and function (package URL from packages management service). |

> | py | N/A| String | `--py` | The path of the main Python/Python wheel file for a function (if the function is written in Python). It also supports URL paths that workers can download the package from, including HTTP, HTTPS, file (file protocol assuming that file already exists on worker host), and function (package URL from packages management service).  |

Reviewer: is ` (if the function is written in Python)` redundant because it already says `... main Python/Python wheel file ...`?

> | go | N/A| String | `--go` | Path to the main Go executable binary for the function (if the function is written in Go).  It also supports URL paths that workers can download the package from, including HTTP, HTTPS, file (file protocol assuming that file already exists on worker host), and function (package URL from packages management service). |

Reviewer: same question as above

> | cleanupSubscription | N/A| Boolean | N/A | Whether the subscriptions that a function created or used should be deleted or not when the function is deleted. |

Reviewer: ```suggestion
| cleanupSubscription | N/A| Boolean | N/A | Whether the subscriptions that a function creates or uses should be deleted or not when the function is deleted. |
```

| customRuntimeOptions | N/A| String | `--custom-runtime-options` | A string that encodes options to customize the runtime. |
| maxPendingAsyncRequests | N/A| Int | `--max-message-retries` | The max number of pending async requests per instance to avoid a large number of concurrent requests. |

> | exposePulsarAdminClientEnabled | N/A| Boolean | N/A | Whether the pulsar admin client is exposed to function context or not. By default, it is disabled. |

Reviewer: ```suggestion
| exposePulsarAdminClientEnabled | N/A| Boolean | N/A | Whether the Pulsar admin client is exposed to function context or not. By default, it is disabled. |
```
